### PR TITLE
Single cilium binary with aliases to optimize container image size

### PR DESCRIPTION
**Pull request by nebril**
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/6125)

### CODE REVIEW

**Comment by nebril** [comment]
test-me-please

**Comment by aanm** [comment]
@nebril the travis failure will be fixed by https://github.com/cilium/cilium/pull/6127

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

File: daemon/main.go [hunk 1]
```diff
@@ -15,1098 +15,24 @@
...
-	if defaultDockerEndpoint != dockerEndpoint {
-		containerRuntimesOpts[string(workloads.Docker)] = dockerEndpoint
-		log.Warn(`"docker" flag is deprecated.` +
-			`Please use "--container-runtime-endpoint=docker=` + defaultDockerEndpoint + `" instead`)
-	}
-
-	err = workloads.ParseConfigEndpoint(option.Config.Workloads, containerRuntimesOpts)
-	if err != nil {
-		log.WithError(err).Fatal("Unable to initialize policy container runtimes")
+	if strings.HasSuffix(os.Args[0], "cilium-agent") {
```
Comment (tgraf) [inline daemon/main.go]: Let's make this more resilient and use `path.Base()` and a string match. `HasSuffix()` can be a bit brittle.

**Review by tgraf — CHANGES_REQUESTED** [review]

**Comment by nebril** [comment]
test-me-please

File: daemon/Makefile [hunk 2]
```diff
@@ -19,25 +20,31 @@ GO_BINDATA := $(QUIET) go-bindata -prefix ../ -mode 0640 -modtime 1450269211 \
...
 	$(QUIET)$(GO) clean
+	$(foreach link,$(LINKS), rm $(link);)
 
 ifeq ("$(PKG_BUILD)","")
 
 install:
 	groupadd -f cilium
 	$(INSTALL) -m 0755 -d $(DESTDIR)$(BINDIR)
 	$(INSTALL) -m 0755 $(TARGET) $(DESTDIR)$(BINDIR)
+	$(foreach link,$(LINKS), ln -s $(TARGET) $(DESTDIR)$(BINDIR)/$(link);)
```
Comment (aanm) [inline daemon/Makefile]: What is this doing?

File: daemon/daemon_main.go [hunk 3]
```diff
@@ -0,0 +1,1113 @@
+// Copyright 2016-2018 Authors of Cilium
```
Comment (aanm) [inline daemon/daemon_main.go:1]: If this file was created, isn't a `daemon/main.go` missing? :thinking:

**Review by aanm — CHANGES_REQUESTED** [review]

File: daemon/Makefile (hunk 2 above)
Comment (nebril) [reply daemon/Makefile]: creating relative symlinks to cilium-agent binary in /usr/bin

File: daemon/daemon_main.go (hunk 3 above)
Comment (nebril) [reply daemon/daemon_main.go:1]: new `daemon/main.go` is in this PR

**Comment by nebril** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

File: daemon/main.go [hunk 4]
```diff
@@ -15,1099 +15,25 @@
...
-		containerRuntimesOpts[string(workloads.Docker)] = dockerEndpoint
-		log.Warn(`"docker" flag is deprecated.` +
-			`Please use "--container-runtime-endpoint=docker=` + defaultDockerEndpoint + `" instead`)
-	}
-
-	err = workloads.ParseConfigEndpoint(option.Config.Workloads, containerRuntimesOpts)
-	if err != nil {
-		log.WithError(err).Fatal("Unable to initialize policy container runtimes")
+	base := path.Base(os.Args[0])
+	if base == "cilium-agent" {
```
Comment (tgraf) [inline daemon/main.go]: This should really be a switch statement

**Review by tgraf — CHANGES_REQUESTED** [review]

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Review by aanm — APPROVED** [review]

**Review by tgraf — APPROVED** [review]

**Pull request merged**